source/framework/vx_array.c

```diff
@@ -22,6 +22,182 @@ static void ownInitArrayObject(
     vx_array arr, vx_enum item_type, vx_size capacity, vx_bool is_virtual);
 static vx_size ownGetArrayItemSize(vx_context context, vx_enum item_type);
 static vx_bool ownIsValidArrayItemType(vx_context context, vx_enum item_type);
+static vx_bool ownIsValidInputAndOutputArrays(vx_array input, vx_array output);
+static vx_status isArrayCopyable(vx_array input, vx_array output);
+static vx_status isArraySwapable(vx_array input, vx_array output);
+static vx_status copyArray(vx_array input, vx_array output);
+static vx_status swapArray(vx_array input, vx_array output);
+static vx_status VX_CALLBACK arrayKernelCallback(vx_enum kernel_enum, vx_bool validate_only, vx_enum optimization, const vx_reference params[], vx_uint32 num_params);
+
+/*! \brief checks that the input and output references are not the same, and that both are valid arrays
+*/
+static vx_bool ownIsValidInputAndOutputArrays(vx_array input, vx_array output)
+{
+    if ((input != output) &&
+        (ownIsValidSpecificReference(&input->base, (vx_enum)VX_TYPE_ARRAY) == (vx_bool)vx_true_e) &&
+        (input->base.obj_desc != NULL) &&
+        (ownIsValidSpecificReference(&output->base, (vx_enum)VX_TYPE_ARRAY) == (vx_bool)vx_true_e) &&
+        (output->base.obj_desc != NULL)
+        )
+    {
+        return vx_true_e;
+    }
+    else
+    {
+        return vx_false_e;
+    }
+}
+
+/*! \brief This function is called to find out if it is OK to copy the input to the output.
+ * If the output is virtual, a zero capacity or invalid item type is OK
+ * Will also copy metadata from input to output.
+ * \returns VX_SUCCESS if it is, otherwise another error code.
+ *
+ */
+static vx_status isArrayCopyable(vx_array input, vx_array output)
```

Comment:
GENERIC: If you want to compare meta format between two references of the same type, we already have a generic function for this: tivxIsReferenceMetaFormatEqual.  You can use this to simplify and generalize for all/most data types.

Reply:
tivxIsReferenceMetaFormatEqual
vx_lut copyable --> ok
vx_matrix --> origin and pattern is missing
vx_convolution --> mo meta comparision for convolution
vx_distrbution --> ok
vx_scalar --> ok
vx_tensor --> not all data checked like scaling_divisor_fixed_point_position
vx_pyramid --> specific checks if it is virtual
vx_image --> also too specific
vx_objarray --> ok, but questions in the code

swapable only two objects have a specific funtion
- vx_image --> cannot be use because of the subimage checked
- vx_array --> ok

copy function: (ownCopyReferenceGeneric)
vx_distrib: done
vx_array : done
vx_matrix: done
vx_lut: done
scalar: not needed
vx_tensor: done
vx_convolution: done
vx_distrbution: done
vx_pyramid: not applicable
vx_image / raw_image : not applicable
vx_objarray: not applicable

swap function:
vx_lut: done
vx_distrbution: done
vx_matrix: done
vx_array: done
vx:tensor: mixed with move, not possible
vx_convolution: done
vx_pyramid: not applicable
vx_image / raw_image : not applicable
vx_objarray: not applicable

Reply:
For your review on tivxIsReferenceMetaFormatEqual, it looks like you identified bugs in missing checks for: 
vx_matrix , vx_convolution , vx_tensor .  If we fixed these, it should make your usage of the tivxIsReferenceMetaFormatEqual more generic I presume.  Please confirm and we can file bugs for this.

Reply:
If I now use the tivxIsReferenceMetaFormatEqual for the vx_matrix , vx_convolution etc., I am not sure if I will pass the test if your fix is not merged. I can try if you want.